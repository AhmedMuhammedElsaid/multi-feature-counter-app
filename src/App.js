import React, { Component, createRef } from "react";
import { connect } from "react-redux";
import { mapDispatchToProps } from "./containers/Controls";
import Counter from "./containers/Counter";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: 1,
    };
    this.incrementRef = createRef();
    this.decrementRef = createRef();
  }
  render() {
    const { increment, decrement } = this.props;
    // control the change of the slider to affects our counter based on the scenario given
    const handleChange = (event) => {
      const newValue = +event.target.value || 1;
      this.setState({ value: newValue });
      increment({
        counterIndx: "counterOne",
        payload: newValue,
      });
      decrement({
        counterIndx: "counterTwo",
        payload: newValue,
      });
    };

    // increase our left counter by 2 every single two seconds

    const handleIncrementToggle = (event) => {
      const { checked } = event.target;
      if (checked) {
        this.incrementRef.current = setInterval(
          () => increment({ counterIndx: "counterOne", payload: 2 }),
          1000
        );
      } else clearInterval(this.incrementRef.current);
    };
    // decrement our left counter by 2 every single two seconds
    const handleDecrementToggle = (event) => {
      const { checked } = event.target;
      if (checked) {
        this.decrementRef.current = setInterval(
          () => decrement({ counterIndx: "counterOne", payload: 2 }),
          1000
        );
      } else clearInterval(this.decrementRef.current);
    };
    return (
      <div className="container">
        <div className="slider-holder">
          <input
            type="range"
            min="1"
            max="100"
            data-testid="slider-testid"
            onChange={handleChange}
            value={this.state.value}
          />
          <p>
            Current Value of the slider is <strong>{this.state.value}</strong>
          </p>
          <div className="toggles-holder">
            <div>
              <p>increment toggle</p>
              <label className="switch">
                <input type="checkbox" onChange={handleIncrementToggle} />
                <span className="slider round"></span>
              </label>
            </div>
            <div>
              <p>decrement toggle</p>
              <label className="switch">
                <input type="checkbox" onChange={handleDecrementToggle} />
                <span className="slider round"></span>
              </label>
            </div>
          </div>
        </div>
        <div className="counters-holder">
          <Counter counterIndx={"counterOne"} />
          <Counter counterIndx={"counterTwo"} disabled={true} />
        </div>
      </div>
    );
  }
}
export default connect(null, mapDispatchToProps)(App);
