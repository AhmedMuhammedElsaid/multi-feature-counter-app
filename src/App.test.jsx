import React from "react";
import "@testing-library/jest-dom";

import { cleanup, render, fireEvent } from "@testing-library/react";
import App from "./App";
import { Provider } from "react-redux";
import { createStore } from "redux";
import counterReducer from "./reducers/counter";
import { combineReducers } from "redux";

const addBtn = "add-btn-testid",
  decBtn = "dec-btn-testid",
  resetBtn = "reset-btn-testid",
  countHolder = "count-value-testid",
  slider = "slider-testid";
  
// this is a fancy function to keep  our component under redux context umbrella
const customRender = (component) => {
  return {
    ...render(
      <Provider
        store={createStore(combineReducers({ counters: counterReducer }))}
      >
        {component}
      </Provider>
    ),
  };
};
describe("test app functionalities", () => {
  afterEach(cleanup);
  it("should render app  without crashing", () => {
    const { asFragment } = customRender(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should change both counters  correctly without crashing", () => {
    const { asFragment, getByTestId, getAllByTestId } = customRender(<App />);
    fireEvent.change(getByTestId(slider), { target: { value: 2 } });
    expect(getAllByTestId(countHolder)[0]).toHaveTextContent(2);
    expect(getAllByTestId(countHolder)[1]).toHaveTextContent(-2);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should increment correctly without crashing", () => {
    const { asFragment, getAllByTestId } = customRender(<App />);
    fireEvent.click(getAllByTestId(addBtn)[0]);
    expect(getAllByTestId(countHolder)[0]).toHaveTextContent(1);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should decrement correctly without crashing", () => {
    const { asFragment, getAllByTestId } = customRender(<App />);
    fireEvent.click(getAllByTestId(decBtn)[0]);
    expect(getAllByTestId(countHolder)[0]).toHaveTextContent(-1);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should reset correctly without crashing", () => {
    const { asFragment, getAllByTestId } = customRender(<App />);
    fireEvent.click(getAllByTestId(resetBtn)[0]);
    expect(getAllByTestId(countHolder)[0]).toHaveTextContent(0);
    expect(asFragment()).toMatchSnapshot();
  });
});
