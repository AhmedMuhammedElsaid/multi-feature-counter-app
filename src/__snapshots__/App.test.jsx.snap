// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test app functionalities should change both counters  correctly without crashing 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="slider-holder"
    >
      <input
        data-testid="slider-testid"
        max="100"
        min="1"
        type="range"
        value="2"
      />
      <p>
        Current Value of the slider is 
        <strong>
          2
        </strong>
      </p>
      <div
        class="toggles-holder"
      >
        <div>
          <p>
            increment toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
        <div>
          <p>
            decrement toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
      </div>
    </div>
    <div
      class="counters-holder"
    >
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          2
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn "
            data-testid="add-btn-testid"
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn "
            data-testid="dec-btn-testid"
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn "
            data-testid="reset-btn-testid"
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          -2
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn disabled"
            data-testid="add-btn-testid"
            disabled=""
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn disabled"
            data-testid="dec-btn-testid"
            disabled=""
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn disabled"
            data-testid="reset-btn-testid"
            disabled=""
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test app functionalities should decrement correctly without crashing 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="slider-holder"
    >
      <input
        data-testid="slider-testid"
        max="100"
        min="1"
        type="range"
        value="1"
      />
      <p>
        Current Value of the slider is 
        <strong>
          1
        </strong>
      </p>
      <div
        class="toggles-holder"
      >
        <div>
          <p>
            increment toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
        <div>
          <p>
            decrement toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
      </div>
    </div>
    <div
      class="counters-holder"
    >
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          -1
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn "
            data-testid="add-btn-testid"
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn "
            data-testid="dec-btn-testid"
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn "
            data-testid="reset-btn-testid"
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          0
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn disabled"
            data-testid="add-btn-testid"
            disabled=""
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn disabled"
            data-testid="dec-btn-testid"
            disabled=""
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn disabled"
            data-testid="reset-btn-testid"
            disabled=""
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test app functionalities should increment correctly without crashing 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="slider-holder"
    >
      <input
        data-testid="slider-testid"
        max="100"
        min="1"
        type="range"
        value="1"
      />
      <p>
        Current Value of the slider is 
        <strong>
          1
        </strong>
      </p>
      <div
        class="toggles-holder"
      >
        <div>
          <p>
            increment toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
        <div>
          <p>
            decrement toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
      </div>
    </div>
    <div
      class="counters-holder"
    >
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          1
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn "
            data-testid="add-btn-testid"
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn "
            data-testid="dec-btn-testid"
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn "
            data-testid="reset-btn-testid"
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          0
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn disabled"
            data-testid="add-btn-testid"
            disabled=""
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn disabled"
            data-testid="dec-btn-testid"
            disabled=""
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn disabled"
            data-testid="reset-btn-testid"
            disabled=""
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test app functionalities should render app  without crashing 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="slider-holder"
    >
      <input
        data-testid="slider-testid"
        max="100"
        min="1"
        type="range"
        value="1"
      />
      <p>
        Current Value of the slider is 
        <strong>
          1
        </strong>
      </p>
      <div
        class="toggles-holder"
      >
        <div>
          <p>
            increment toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
        <div>
          <p>
            decrement toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
      </div>
    </div>
    <div
      class="counters-holder"
    >
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          0
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn "
            data-testid="add-btn-testid"
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn "
            data-testid="dec-btn-testid"
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn "
            data-testid="reset-btn-testid"
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          0
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn disabled"
            data-testid="add-btn-testid"
            disabled=""
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn disabled"
            data-testid="dec-btn-testid"
            disabled=""
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn disabled"
            data-testid="reset-btn-testid"
            disabled=""
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test app functionalities should reset correctly without crashing 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="slider-holder"
    >
      <input
        data-testid="slider-testid"
        max="100"
        min="1"
        type="range"
        value="1"
      />
      <p>
        Current Value of the slider is 
        <strong>
          1
        </strong>
      </p>
      <div
        class="toggles-holder"
      >
        <div>
          <p>
            increment toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
        <div>
          <p>
            decrement toggle
          </p>
          <label
            class="switch"
          >
            <input
              type="checkbox"
            />
            <span
              class="slider round"
            />
          </label>
        </div>
      </div>
    </div>
    <div
      class="counters-holder"
    >
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          0
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn "
            data-testid="add-btn-testid"
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn "
            data-testid="dec-btn-testid"
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn "
            data-testid="reset-btn-testid"
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
      <div
        class="counter"
      >
        <div
          data-testid="count-value-testid"
        >
          0
        </div>
        <div
          class="controls"
        >
          <button
            class="addBtn disabled"
            data-testid="add-btn-testid"
            disabled=""
          >
            <span>
              +
            </span>
          </button>
          <button
            class="decBtn disabled"
            data-testid="dec-btn-testid"
            disabled=""
          >
            <span>
              -
            </span>
          </button>
          <button
            class="resetBtn disabled"
            data-testid="reset-btn-testid"
            disabled=""
          >
            <span>
              reset
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
